### PR TITLE
Fixes #390: Split `send_analysis_request` into `analyze_file` and `analyze_observable`

### LINKED ISSUE #390: Divide `/send_analysis_request` endpoint into `/analyse_observable` and `/analyse_file`
**Note: Would also require changes in the pyintelowl and IntelOwl-ng**

On the client side, we already provide two different methods to build the request body differently based on whether user chose `observable` or `file`. So it just makes sense to divide the endpoint into two for easier management and less confusion.

### PR DESCRIPTION
**Pull request by m0mosenpai**
# Description

- [X] New views for endpoints
- [X] New Serializers
- [x] Testing and Refactoring

## Related issues
Fixes #390 

## Type of change

- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected).

# Checklist

- [X] I have added tests in the [Tests](https://github.com/intelowlproject/IntelOwl/blob/master/tests) folder. 
- [x] The tests gave 0 errors.
- [X] `Black` gave 0 errors.
- [X] `Flake` gave 0 errors.

### CODE REVIEW

**Comment by m0mosenpai** [comment]
https://github.com/intelowlproject/IntelOwl/pull/547#issuecomment-882345688

@Eshaan7 
Regarding this, It requires passing the function the `file` and `file_name` arguments. I'm not sure how i can access those values here. ```file_mimetype = serializers.HiddenField(default=calculate_mimetype(file_buffer, file_name))```

File: tests/test_api.py [hunk 1]
```diff
@@ -109,9 +107,7 @@ def test_send_analysis_request_sample(self):
             "file": uploaded_file,
             "test": True,
         }
-        response = self.client.post(
-            "/api/send_analysis_request", data, format="multipart"
-        )
+        response = self.client.post("/api/analyze_file", data, format="multipart")
         self.assertEqual(response.status_code, 200)
 
     def test_send_analysis_request_domain(self):
```
Comment (Eshaan7) [inline tests/test_api.py:113]: also rename these function like `test_analyse_observable_ip` or `test_analyse_file` etc.

**Comment by Eshaan7** [comment]
I am not sure but I think it should be possible to do like:

```python3
file_mimetype = serializers.HiddenField(default=lambda obj: calculate_mimetype(obj.file, obj.file_name))
```

File: api_app/api.py [hunk 2]
```diff
@@ -32,6 +33,77 @@
...
+        "job_id": prep_data["job_id"],
+        "analyzers_running": prep_data["analyzers_to_execute"],
+    }
+
+    logger.debug(response_dict)
+
+    return response_dict
+
+
+def prepare_for_analysis(serializer, params, warnings):
```
Comment (Eshaan7) [inline api_app/api.py:64]: While this approach with 2 common functions `send_for_analysis` and `prepare_for_anlaysis` is okay, there's a better way I just thought of:

Combine these 2 functions into an individual function 

```python3
def _analysis_request(request, serializer_class):
```

And then the `analyze_file` and `analyze_observable` would look like:

```python3
@api_view()
def analyze_file(request):
    return _analysis_request(request, FileAnalysisSerializer)

@api_view()
def analyze_observable(request):
    return _analysis_request(request, ObservableAnalysisSerializer)
```

**Comment by m0mosenpai** [comment]
> I am not sure but I think it should be possible to do like:
> 
> ```python
> file_mimetype = serializers.HiddenField(default=lambda attrs: calculate_mimetype(attrs["file"], attrs["file_name"]))
> ```

I don't think this works. It says it's missing a positional requirement `attrs`. I tried looking it up online as well but didn't find examples where they had called a function with positional arguments in `HiddenField`

**Comment by Eshaan7** [comment]
> I don't think this works. It says it's missing a positional argument `attrs`. I tried looking it up online as well but didn't find examples where they had called a function with positional arguments in `HiddenField`

Okay, we need to have a hacky implementation then:
- keep the `SerializerMethodField` just the way it is right now.
- overwrite `validate(self, attrs)` function of `FileAnalysisSerializer` like:

```python3
def validate(self, attrs):
  super().validate(attrs)
  attrs["file_mimetype"] = self.get_file_mimetype(attrs)
```

File: api_app/serializers.py [hunk 3]
```diff
@@ -81,16 +83,10 @@ def get_no_of_connectors_executed(self, obj: Job):
 class JobSerializer(ObjectPermissionsAssignmentMixin, serializers.ModelSerializer):
```
Comment (Eshaan7) [inline api_app/serializers.py:83]: `ObjectPermissionsAssignmentMixin` and `get_permissions_map` should be moved to the 2 new serializers as well.

Comment (Eshaan7) [reply api_app/serializers.py:83]: Create a base serializer like `class JobCreateSerializerMixin(ObjectPermissionsAssignmentMixin)` to keep the common code like `validate`, `create`, `get_permissions_map` methods.

**Comment by Eshaan7** [comment]
Looks good to me. Can you run the relevant tests on your local to test and post the screenshot here ? (inconvenience caused since the testing suite is broken for develop-2 branch)

**Comment by m0mosenpai** [comment]
API Tests:
image

GUI:
Observable:
image

FIle:
image

File: api_app/api.py [hunk 4]
```diff
@@ -199,6 +287,95 @@ def ask_analysis_availability(request):
...
+            description=r"""
+            Default {}, contains additional parameters for particular analyzers""",
+        ),
+        OpenApiParameter(
+            name="test",
+            type=OpenApiTypes.BOOL,
+            description="disable analysis for API testing",
+        ),
+    ],
+    responses={
```
Comment (Eshaan7) [inline api_app/api.py:337]: We do not need to specify the parameters now, set this like: `request=serializers.FileAnalysisSerializer` (refer to: vhttps://drf-spectacular.readthedocs.io/en/latest/customization.html#step-2-extend-schema)

Comment (Eshaan7) [reply api_app/api.py:337]: same for the `analyze_observable`.

File: api_app/serializers.py [hunk 5]
```diff
@@ -120,12 +116,57 @@ def get_permissions_map(self, created):
...
+                raise serializers.ValidationError(
+                    {
+                        """
+                        analyzers_requested and run_all_available_analyzers
+                        cannot be used together
+                        """
+                    }
+                )
+
+            data["analyzers_requested"] = analyzers_controller.ALL_ANALYZERS
```
Comment (Eshaan7) [inline api_app/serializers.py:168]: Remove this line. And then in the `_analysis_request` function check the `if` condition again for this. (as was done before)

File: api_app/serializers.py [hunk 6]
```diff
@@ -135,3 +181,81 @@ def create(self, validated_data):
...
+    Job model's serializer for Observable Analysis.
+    Used for create()
+    """
+
+    tags_id = serializers.PrimaryKeyRelatedField(
+        many=True, write_only=True, queryset=Tag.objects.all()
+    )
+    runtime_configuration = serializers.JSONField(
+        required=False, default={}, write_only=True
+    )
```
Comment (Eshaan7) [inline api_app/serializers.py:240]: Is it not possible to move these common fields to the `_JobCreateSerializerMixin` ?

**Pull request merged**